src/modules/aws_api_gateway/mappers/api.ts

```diff
@@ -0,0 +1,199 @@
+        }
+        const newApi: Api = await this.cloud.read(
+          ctx,
+          this.generateId({ apiId: result?.ApiId ?? '', region: r.region }),
+        );
+        if (!newApi) continue;
+        // use the same ID as the one inserted, and set the name as is optionally returned
+        newApi.id = r.id;
+        newApi.name = r.name;
+        await this.module.api.db.update(newApi, ctx);
+        out.push(newApi);
+      }
+      return out;
+    },
+    read: async (ctx: Context, id?: string) => {
+      const enabledRegions = (await ctx.getEnabledAwsRegions()) as string[];
+      if (!!id) {
+        const { apiId, region } = this.idFields(id);
+        if (enabledRegions.includes(region)) {
+          const client = (await ctx.getAwsClient(region)) as AWS;
+          const rawApi = await this.getApi(client.apiGatewayClient, apiId);
+          if (rawApi) {
+            return this.apiMapper(rawApi, region);
+          }
+        }
+      } else {
+        const out: Api[] = [];
+        await Promise.all(
+          enabledRegions.map(async region => {
+            const client = (await ctx.getAwsClient(region)) as AWS;
+            const rawApis = (await this.getApis(client.apiGatewayClient)) ?? [];
+            for (const i of rawApis) {
+              const outApi = this.apiMapper(i, region);
+              if (outApi) out.push(outApi);
+            }
+          }),
+        );
+        return out;
+      }
+    },
+    update: async (rs: Api[], ctx: Context) => {
+      const out: Api[] = [];
+      for (const r of rs) {
+        const client = (await ctx.getAwsClient(r.region)) as AWS;
+        const cloudRecord = ctx?.memo?.cloud?.Api?.[this.entityId(r)];
+        // api id is generated by aws, we cannot modify it
+        if (cloudRecord.protocolType !== r.protocolType) {
+          // restore
+          cloudRecord.id = r.id;
+          await this.module.api.db.update(cloudRecord, ctx);
+          out.push(cloudRecord);
+        } else {
+          const input: UpdateApiCommandInput = {
+            ApiId: r.apiId,
+            Name: r.name,
+            Description: r.description,
+            DisableExecuteApiEndpoint: r.disableExecuteApiEndpoint,
+            Version: r.version,
+          };
+          const res = await this.updateApi(client.apiGatewayClient, input);
+          if (!res) continue;
+          if (!eqTags(r.tags, cloudRecord.tags)) {
+            const apiArn = this.buildApiArn(r.apiId, r.region);
+            await this.untagRes(client.apiGatewayClient, {
+              ResourceArn: apiArn,
+              TagKeys: Object.keys(cloudRecord.tags ?? {}),
+            });
+            // sleep for a few seconds to allow the tags to be applied
+            await new Promise(resolve => setTimeout(resolve, 5000));
+            await this.tagRes(client.apiGatewayClient, { ResourceArn: apiArn, Tags: r.tags });
+            // sleep for a few seconds to allow the tags to be applied
```

Comment:
Is there a way with API calls to do this? I know tags should not take more than 5 seconds to apply, but if there's a way to wait for the result in a more programmatic way I think it'd be better.

Reply:
yeah, agree it is definitely not ideal but already spent too much time on this change. There were a lot of inconsistencies with this API. I can try to reduce the time to a couple of seconds and see if tests still pass and the programmatic version can come as a follow-up.

Reply:
The AWS API caches based on the exact input for 40 seconds. It might be the case since I had that issue with the inconsistencies: https://gist.github.com/mtp1376/478860e7acbe55fab3727db968375eaf

Reply:
Addressed using waiters!!